Add Joint Distribution

nabriis opened the pull request:
Closes #21 

- [x] Add Joint distribution
- [x] Add Docs
- [x] Add tests
- [x] Add Stacked Joint Distribution
- [x] Add tests for stacked
- [x] Add  tutorial

The tutorial file file can be compiled to ipynb same way we compile docs
```
sphinx-build -b html docs/. docs/_build
```
and overview with links will be under _build\user\_auto_tutorials\JointDistribution.html in cuqipy directory

nabriis commented:
Hi @amal-ghamdi. I added you as initial reviewer on this PR.

jakobsj reviewed (commented):
Being a distribution, should joint distribution have a name?  At the moment `joint.name` fails since it does not a have a name.  This relates to the discussion of whether "name" or "random variable" etc should be used. It is not clear how one might define a "name" for joint based on the names of the densities "x" and "y" it is constructed from, but it seems okay to let it have a list of random variables ["x", "y"] similarly to geometry and dim.

jakobsj commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+# dim (dimension) and geometry. Because its a joint distribution
+# over multiple variables, these are lists.
+
+print(joint.dim)
```
The dimension of y is None, is that intended? Being defined through a lambda function its size is perhaps hard to know in general?

jakobsj replied:
I think we agreed you would create an issue on allowing the user to specify dimension manually?

nabriis replied:
#28

jakobsj commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+y = GaussianCov(lambda x: A@x, s*Im)
+
+# Define joint distribution p(x,y)
+joint = JointDistribution([x, y])
```
We construct the joint from the two parameters x and y. But `joint.get_parameter_names()` returns the four parameters
`['d', 'l', 'x', 'y']`  and also prints the four parameter
```

print(joint)
JointDistribution(
    Equation: 
	p(d,l,x,y) = p(d)p(l)p(x|d)p(y|x,l)
    Densities: 
	d ~ CUQI Gamma.
	l ~ CUQI Gamma.
	x ~ CUQI GaussianCov. Conditioning variables ['d'].
	y ~ CUQI GaussianCov. Conditioning variables ['x', 'l'].
)
```
That is not completely intuitive?  We don't need to define the joint from all four?

jakobsj commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+# arguments.
+
+# Using keyword arguments
+logd = joint.logd(y=np.ones(m), x=np.ones(n))
```
Related to the above point on two vs four parameters, here values of the two which completely specifies everything so we can evaluate the logd. However, if we, for whatever reason maybe as mistake, specify just one, e.g., 
```
joint.logd(y=np.ones(m))
```
we get an error that tells us we need to specify all four:
```
ValueError: To evaluate the log density function, all parameters must be passed. Received dict_keys(['y']) and expected ['d', 'l', 'x', 'y'].
```

jakobsj commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+# This returns a new "stacked" joint distribution that the samplers/solvers
+# can use as if it were any other Density.
+
+posterior_stacked = joint(y=y_obs)._as_stacked()
```
```suggestion
posterior_stacked = posterior._as_stacked()
```

jakobsj replied:
Reusing already created `posterior` for clarity and simplicity.

jakobsj reviewed (changes requested):
I went through the demo again and that resolved the issues about two or four parameters, my fault.  Super nice work, thank you for adding the comprehensive demo and for the explanations earlier, that helped a lot in reviewing this. Only added some minor requests, and re `.name` of joints and posteriors and option to equip distributions with dimension manually please create issues. I haven't given much scrutiny to the private helper functions.

nabriis commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+# Because the posterior no longer depends on the data, the dimension is only
+# with respect to the parameters.
+print(f"Posterior dimension: {posterior.dim}")
+print(f"Posterior geometry: {posterior.geometry}")
```
The Posterior class does have the option to get a name attribute similar to any other Distribution. But the "posterior" from the demo is still a joint distribution. Added an issue to rethink this name attribute #29

nabriis commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,233 @@
+y = GaussianCov(lambda x: A@x, lambda l: l*Im)
+
+# Define joint distribution p(d,l,x,y)
+joint = JointDistribution([d, l, x, y])
```
Done. Thanks :)

nabriis commented:
> Being a distribution, should joint distribution have a name? At the moment `joint.name` fails since it does not a have a name. This relates to the discussion of whether "name" or "random variable" etc should be used. It is not clear how one might define a "name" for joint based on the names of the densities "x" and "y" it is constructed from, but it seems okay to let it have a list of random variables ["x", "y"] similarly to geometry and dim.

Hi @jakobsj. Thank you for the review. I incorporated most of your comments. I added two new issues coming from our discussions as you can see in the comments. Finally after our retro I also split the demo file for the joint distribution into a tutorial relevant to users and a developer demo. I added the developer demo in the contributor guide under a new section.

jakobsj reviewed (approved):
Good idea to split off a technical demo and starting a collection of such demos.

amal-ghamdi commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,175 @@
+#
+# .. math::
+#
+#   \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#   \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
same comment about d and s here (as in dev/JointDistribution.py)

amal-ghamdi commented on demos/tutorials/JointDistribution.py:
```diff
@@ -0,0 +1,175 @@
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1,10^{-4}) \\
+#         \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
same comment about d and s here (as in dev/JointDistribution.py)

amal-ghamdi reviewed (commented):
Thank you @nabriis for adding these great features! The PR is very nice. I had few minor comments, mostly about documentation.

nabriis commented on cuqi/distribution/_joint_distribution.py:
```diff
@@ -0,0 +1,302 @@
+        return kwargs
+
+    def _reduce_to_single_density(self):
+        """ Reduce the joint distribution to a single density if possible.
```
Good suggestion. Thanks

nabriis commented on tests/test_joint_distribution.py:
```diff
@@ -0,0 +1,245 @@
+    dims = J.dim
+
+    # Make dictionary of random parameters matching names and dimensions
+    params = {name: abs(np.random.randn(dim)) for name, dim in zip(names, dims)}
```
Hmm. The random number was part of the test. If it starts randomly failing for no good reason we can consider adding a seed.

nabriis commented on demos/dev/JointDistribution.py:
```diff
@@ -0,0 +1,135 @@
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1,10^{-4}) \\
+#         \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
Good catch. Was a typo. I updated it to d^{-1} and l^{-1} to match code

nabriis commented on demos/dev/JointDistribution.py:
```diff
@@ -0,0 +1,135 @@
+#       C_d&=p(d \mid \hat{\mathbf{y}}, \hat{\mathbf{x}}, \hat{l})\\
+#       C_l&=p(l \mid \hat{\mathbf{y}}, \hat{\mathbf{x}}, \hat{d})
+#
+# Assuming we have some values for :math:`\mathbf{x}`, :math:`d` and :math:`l`,
```
Added. Thanks

nabriis commented:
Thanks for the review @amal-ghamdi. Good you caught those inconsistencies!

amal-ghamdi reviewed (approved):
Looks great! Thank you @nabriis.
